pkg/counter/counter.go

```diff
@@ -1,31 +1,94 @@
 package counter
 
-import "sync/atomic"
+import (
+	"fmt"
+	"strconv"
+	"sync/atomic"
+)
 
-type Counter int32
+type Counter struct {
+	count uint64
+}
 
-func (c *Counter) Increment(amount ...int) {
-	sum := 1
-	if len(amount) > 0 {
-		sum = 0
-		for _, a := range amount {
-			sum = sum + a
-		}
+func NewCounter(v uint64) Counter {
+	return Counter{
+		count: v,
+	}
+}
+
+func (c *Counter) Increment(amount ...uint64) error {
+	if len(amount) == 0 {
+		return c.incAtomic(1)
+	}
+
+	sum, err := sumUint64(amount...)
+	if err != nil {
+		return err
+	}
+
+	return c.incAtomic(sum)
+}
+
+func (c *Counter) Decrement(amount ...uint64) error {
+	if len(amount) == 0 {
+		return c.decAtomic(1)
+	}
+
+	sum, err := sumUint64(amount...)
+	if err != nil {
+		return err
 	}
-	atomic.AddInt32((*int32)(c), int32(sum))
+
+	return c.decAtomic(sum)
+}
+
+func (c *Counter) Read() uint64 {
+	return atomic.LoadUint64(&c.count)
 }
 
-func (c *Counter) Decrement(amount ...int) {
-	sum := -1
-	if len(amount) > 0 {
-		sum = 0
-		for _, a := range amount {
-			sum = sum - a
+func (c *Counter) Set(val uint64) {
+	atomic.StoreUint64(&c.count, val)
+}
+
+func (c Counter) Format(f fmt.State, r rune) {
+	f.Write([]byte(strconv.FormatUint(c.count, 10)))
+}
+
+func (c *Counter) incAtomic(v uint64) error {
+	if (v + c.count) < c.count {
+		return errorCounterWrapAround()
+	}
+	atomic.AddUint64(&c.count, v)
+
+	return nil
+}
+
+func (c *Counter) decAtomic(v uint64) error {
+	v = ^uint64(v - 1)
+	if (v + c.count) > c.count {
+		return errorCounterWrapAround()
+	}
+	atomic.AddUint64(&c.count, v)
```

Comment:
v is of type uint, so I don't understand how this code works (will always increment and not decrement)

Reply:
atomic pkg only provides Add (no Sub). To get around this is just to do a two's complement calculation: v = ^uint64(v - 1).

So the adder will subtract instead.